rsfg/src/lexer.rs

```diff
@@ -2,146 +2,187 @@
 
 use crate::Token;
 
-static SPACE: &str = " \t";
-
-// Rules: A-Z,a-z
-fn is_id_1st(c: char) -> bool {
-	c >= 'A' && c <= 'z'
-}
 // A-Z or 0-9
 fn is_id(c: char) -> bool {
-	is_id_1st(c) || (c >= '0' && c <= '9')
+	match c {
+		'0'..='9' | 'A'..='z' => true,
+		_ => false
+	}
+}
+
+enum NextTokenType {
+	EOF,
+	SymbolOrId(char),
+	Space(char),
+	Digit(char),
+	CommentOrDivision,
+	Newline,
+	LParen,
+	RParen,
+	Colon,
+	Comma,
+	Quote,
+	Unknown(char),
+}
+
+#[derive(Debug)]
+struct Lexer<'src> {
+	source: &'src str,
+	tokens: Vec<Token>,
+	spaces_count: usize,
+	rchars: Vec<char>,
+}
+
+impl<'src> Lexer<'src> {
+	pub fn new(source: &'src str) -> Lexer<'src> {
+		Lexer {
+			source,
+			tokens: Vec::new(),
+			spaces_count: 0,
+			rchars: source.chars().rev().collect(),
+		}
+	}
+
+	fn next_symbol_type(&mut self) -> NextTokenType {
+		use NextTokenType::*;
+
+		match self.rchars.pop() {
+			None => EOF,
+			Some(c) => match c {
+				'A'..='z' => SymbolOrId(c),
+				' ' | '\t' => Space(c),
+				'0'..='9' => Digit(c),
+				'/' => CommentOrDivision,
+				'\n' => Newline,
+				'(' => LParen,
+				')' => RParen,
+				':' => Colon,
+				',' => Comma,
+				'"' => Quote,
+				o => Unknown(o)
+			}
+		}
+	}
 }
 
 pub fn lex(text: &str) -> Vec<Token> {
 	use Token::*;
-	// Start with a newline. We could probly find a better way but meh
-	let mut tokens = Vec::<Token>::new();
-	let mut spaces_count = 0;
-	let mut rchars: Vec<char> = text.chars().collect();
-	rchars.reverse();
+	let mut lexer = Lexer::new(text);
 	loop {
-		let c = match rchars.last() {
-			Some(c) => *c,
-			None => break, // We weren't in the middle of anything so successful EOF
-		};
-		let token = if is_id_1st(c) {
-			let mut text = c.to_string();
-			rchars.pop();
-			loop {
-				let x = match rchars.last() {
-					Some(x) => *x,
-					None => break, // End of ID is fine
-				};
-				if is_id(x) {
-					text.push(x);
-				} else {
-					break;
-				}
-				rchars.pop();
+		println!("{:?}", lexer);
```

Comment:
oh okay, i see this is a demonstration of how bundling things into a struct makes debugging easier. i like it! although it's only 4 variables.... either way this should come out when we're done

Reply:
Yeah, definitely.